### ISSUE TITLE
response is empty in browser.getRequests() array

### ISSUE THREAD

**Issue by navi-kris** [webdriverio-community/wdio-intercept-service#204]
Hi I am trying to use WDIO intercept service to get the response for a POST call made in UI. I want to get the response for 

Test:
```
describe('test', () => {
    it('test', () => {
        browser.url('<application url>')
        browser.pause(2000);
        actions.setValueAndHitEnter($('#txtInputTechnicianId'), '123456', 'Tech ID');
        actions.setValueAndHitEnter($('#singleOrder'), '1030703234', ' id');
        actions.setValueAndHitEnter($('//input[@name="serviceTypeName"]'), 'Priority Overnight', 'service type');
        browser.setupInterceptor();
        actions.click($('#btnRePrintLbl'), 'Print button');
        actions.waiting(10);
        var result = browser.getRequests();
        console.log('RESULT***' + JSON.stringify(result));
    })
})
```

I am trying to get all the requests made after clicking print button. I am getting the request details. But the response is empty for all the requests. Particularly I want the response for a call made after clicking the button.

This is the response for all calls

```
        "response": {
            "headers": {},
            "body": "",
            "statusCode": 0
        }
```
I am using  "wdio-intercept-service": "^4.2.1",

**Comment by tehhowch** [comment]
If those requests are not yet completed, just calling `browser.getRequests()` will not return them. What shows up if you use `browser.getRequests({ includePending: true })` instead?

**Comment by navi-kris** [comment]
> If those requests are not yet completed, just calling `browser.getRequests()` will not return them. What shows up if you use `browser.getRequests({ includePending: true })` instead?
> 
> Also, a statusCode of `0` implies some issue with your tested application.
> 
> Can you share more about how the application is making the requests?

When I tried with `browser.getRequests({ includePending: true })` I see the same response for this request. But additionally I got one more request which was in pending state earlier.

This application is used for printing shipping label after scanning the product. So after filling all required details, clicking print button will invoke first call - to generate shipment details. After few seconds second call will be made automatically and then two more requests. So 4 requests in total. The first one will be after clicking on print button and rest will follow based on previous calls response.

**Comment by tehhowch** [comment]
I mean _specifically_. Fetch? XmlHttpRequest? What protocol? How is the url set up (e.g. string, a `Url`, a `Request`, etc.)? What content-type of response did the server send? What cors modes are used? Etc.

**Comment by navi-kris** [comment]
@tehhowch I am not a developer and don't know the technical details of the request. But I have attached the call details from the network tab. Will this help?

![image](https://user-images.githubusercontent.com/32604166/149389837-afbd841c-c96a-4bab-9ea9-1159132b7e45.png)

**Comment by tehhowch** [comment]
> @tehhowch I am not a developer and don't know the technical details of the request.

Ask the developers on your team for this information. If they won't give it to you, escalate with your product owner. Someone you work with should want to make sure their product can be tested properly

**Comment by lildesert** [comment]
Hi there, I'm having the same issue with version `4.2.1`. After digging a bit I discovered the issue was introduced by this PR https://github.com/webdriverio-community/wdio-intercept-service/pull/129. (So the quickest fix would be to use version `4.1.3` but that's not future proof)

I don't have a good way to reproduce it yet, but the thing is the `_this` reference [here](https://github.com/webdriverio-community/wdio-intercept-service/blob/main/lib/interceptor.js#L114) isn't always updated. 

In my scenario, I'm intercepting a `PUT` request, here's my test function:
```
export async function campaignHasInvoicingDisabled() {
  await changeEntityContext(this.entityId)
  await navigateTo(`/edit-campaign/${this.latestCampaignId}`)
  await setValueOfInput('[data-elem="budget-input"]', '200')
  browser.setupInterceptor()
  await $('[data-elem="submit-btn"]').click()
  await browser.pause(1000)
  const requests = await browser.getRequests()
  const updateCampaignRequest = requests.find(r => r.url.includes('api/campaign'))
  console.log(
    '🚀 ~ file: autoInvoicing.ts ~ line 31 ~ campaignHasAutoInvoicingDisabled ~ updateCampaignRequest',
    updateCampaignRequest
  )
  assert.equal(updateCampaignRequest?.response.headers['disable-auto-invoicing'], 'ok')
}
```

Which gives me the same response content
```
response: { headers: {}, body: '', statusCode: 0 }
```

I'm still trying to find a fix that covers every use case.

**Comment by tehhowch** [comment]
> Hi there, I'm having the same issue with version `4.2.1`. After digging a bit I discovered the issue was introduced by this PR https://github.com/webdriverio-community/wdio-intercept-service/pull/129. (So the quickest fix would be to use version `4.1.3` but that's not future proof)

Awesome that you traced the issue source, @lildesert ! 

> I don't have a good way to reproduce it yet, but the thing is the `_this` reference [here](https://github.com/webdriverio-community/wdio-intercept-service/blob/main/lib/interceptor.js#L114) isn't always updated. 

I'm not opposed to this library implementing more robust integration tests that run against a variety of popular frontend SPAs (Angular, Ember, etc.); right now the library tests against a simple static server with no-frills webpages.

In other words, if you can write a failing test case for this library using a stripped-down angular app, I think that'd be awesome.


> In my scenario, I'm intercepting a `PUT` request, here's my test function:
> ```
> export async function campaignHasInvoicingDisabled() {
>   await changeEntityContext(this.entityId)
>   await navigateTo(`/edit-campaign/${this.latestCampaignId}`)
>   await setValueOfInput('[data-elem="budget-input"]', '200')
>   browser.setupInterceptor()
>   await $('[data-elem="submit-btn"]').click()
>   await browser.pause(1000)
>   const requests = await browser.getRequests()
>   const updateCampaignRequest = requests.find(r => r.url.includes('api/campaign'))
>   console.log(
>     '🚀 ~ file: autoInvoicing.ts ~ line 31 ~ campaignHasAutoInvoicingDisabled ~ updateCampaignRequest',
>     updateCampaignRequest
>   )
>   assert.equal(updateCampaignRequest?.response.headers['disable-auto-invoicing'], 'ok')
> }
> ```


This test case has a minor issue - you need to `await` all custom commands in wdio's async mode, but you have called `browser.setupInterceptor();`. I doubt that is the reason for this issue, though.

**Comment by navi-kris** [comment]
> Hi there, I'm having the same issue with version `4.2.1`. After digging a bit I discovered the issue was introduced by this PR #129. (So the quickest fix would be to use version `4.1.3` but that's not future proof)
> 
> I don't have a good way to reproduce it yet, but the thing is the `_this` reference [here](https://github.com/webdriverio-community/wdio-intercept-service/blob/main/lib/interceptor.js#L114) isn't always updated.
> 
> In my scenario, I'm intercepting a `PUT` request, here's my test function:
> 
> ```
> export async function campaignHasInvoicingDisabled() {
>   await changeEntityContext(this.entityId)
>   await navigateTo(`/edit-campaign/${this.latestCampaignId}`)
>   await setValueOfInput('[data-elem="budget-input"]', '200')
>   browser.setupInterceptor()
>   await $('[data-elem="submit-btn"]').click()
>   await browser.pause(1000)
>   const requests = await browser.getRequests()
>   const updateCampaignRequest = requests.find(r => r.url.includes('api/campaign'))
>   console.log(
>     '🚀 ~ file: autoInvoicing.ts ~ line 31 ~ campaignHasAutoInvoicingDisabled ~ updateCampaignRequest',
>     updateCampaignRequest
>   )
>   assert.equal(updateCampaignRequest?.response.headers['disable-auto-invoicing'], 'ok')
> }
> ```
> 
> Which gives me the same response content
> 
> ```
> response: { headers: {}, body: '', statusCode: 0 }
> ```
> 
> I'm still trying to find a fix that covers every use case.
> 
> Edit: to add a bit more context, the client request is made using angular v6 `HttpClient`

@lildesert  Wow! Thank you! I am not seeing this issue with wdio-intercept-service": "4.1.3". I am getting the response details.

**Comment by lildesert** [comment]
So I spent some time writing a test with angular. It worked fine when I first tested with angular `12.2.16`. I tried with several other angular versions until I found the root of the issue (at least in my case) -> https://github.com/angular/angular/issues/36537
It was fixed in angular `10.0.1` by this PR https://github.com/angular/angular/commit/a5d5f67be7407be0e6eb2c059356c23e83049b13

I opened a PR with my test since it can't harm but in my case, I'll pin the wdio-intercept-service version to `4.1.3` until we bump our angular version (currently at `6.x.x`)

If @navi-kris issue is also related to angular or similar behavior in another lib I guess we can close this issue.
